src/Engine/Data/Texture.hpp

```diff
@@ -16,32 +17,32 @@ class Texture;
 namespace Ra {
 namespace Engine {
 namespace Data {
-/**
- * Describes the content and parameters of a texture.
- * This structures encapsulates all the states used for creating an OpenGL texture.
- *  These parameters describe the image data of the texture :
- *    - target, width, height, depth, format, internalFormat, type and texels for describing image
- * data
- *    - wrapS, wrapT, wrapP, minFilter and magFilter for describing the sampler of the texture.
- *
- *  When one wants to create a texture, the first thing to do is to create and fill a Texture
- * parameter structure that will describe the Texture.
- *
- *  The Texture creation could be done either using the TextureManager or directly on the client
- * class/function.
- *
- *  When a texture is created, no OpenGL initialisation is realized. The user must first call
- * initializeGL before being able to use this texture in an OpenGL operation.
- *
- *  MipMap representation of the texture is automatically generated as soon as the minFilter
- * parameter is something else than GL_LINEAR or GL_NEAREST
- *
- * @note No coherence checking will be done on the content of this structure. User must ensure
- * coherent data and parameters before creating the OpenGL texture with Texture::initializeGL
- */
-struct TextureParameters {
-    /// Name of the texture
-    std::string name {};
+
+struct SamplerParameters {
+    /// OpenGL wrap mode in the s direction
+    GLenum wrapS { GL_CLAMP_TO_EDGE };
+    /// OpenGL wrap mode in the t direction
+    GLenum wrapT { GL_CLAMP_TO_EDGE };
+    /// OpenGL wrap mode in the p direction
+    GLenum wrapP { GL_CLAMP_TO_EDGE };
+    /// OpenGL minification filter ( GL_LINEAR or GL_NEAREST or GL_XXX_MIPMAP_YYY )
+    GLenum minFilter { GL_LINEAR };
+    /// OpenGL magnification filter ( GL_LINEAR or GL_NEAREST )
+    GLenum magFilter { GL_LINEAR };
+};
+
+struct ImageParameters {
+
+    using Image2DType = std::shared_ptr<void>;
+    using CubeMapType = std::array<std::shared_ptr<void>, 6>;
+    const Image2DType* getImage2D() const {
+        return std::get_if<ImageParameters::Image2DType>( &texels );
+    }
+
+    const CubeMapType* getCubeMap() const {
+        return std::get_if<ImageParameters::CubeMapType>( &texels );
+    }
```

Comment:
Right, the current verion is a first step (trying to change texture implem, keeping the interface), and will be adapted.
I'll prefer on my side trying to avoid exception as much as possible, but the proposed interface is ok.

Reply:
I'll prefer on my side trying to avoid exception as much as possible, but the proposed interface is ok.
We can use std::optional.
Why the const * to shared is dangerous ?

Reply:
Done, feel free to reopen if needed